src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,355 @@
+        return Nothing
+      Reader.ReadSmallEntry readCtxHeadKey entry -> do
+        -- Note that this small entry could be the only one on the page. We
+        -- only care about it being small, not single-entry, since it could
+        -- still end up sharing a page with other entries in the merged run.
+        -- TODO(optimise): This currently doesn't fully exploit the case where
+        -- there is a single page small entry on the page which again ends up
+        -- as the only entry of a page. In that case we could have copied the
+        -- RawPage (but we find out too late to easily exploit it).
+        let readCtxHeadEntry = FullEntry entry
+        return (Just ReadCtx {..})
+      Reader.ReadLargeEntry readCtxHeadKey entry rawPage lenSuffix overflowPages -> do
+        let overflowEntry = OverflowEntry entry rawPage lenSuffix overflowPages
+        let readCtxHeadEntry = case entry of
+              Mupdate _ -> FullEntry (toFullEntry overflowEntry)
+              _         -> overflowEntry
+        return (Just ReadCtx {..})
+
+-- | Only call when aborting before all readers have been drained.
+closeReaders ::
+     HasFS IO h
+  -> Readers (FS.Handle h)
+  -> IO ()
+closeReaders fs Readers {..} = do
+    (reader, _) <- readIORef readersLastRead
+    Reader.close fs reader
+    go
+  where
+    go = Heap.extract readersHeap >>= \case
+          Nothing -> return ()
+          Just ReadCtx {readCtxReader} -> do
+            Reader.close fs readCtxReader
+            go
+
+{-------------------------------------------------------------------------------
+  Write
+-------------------------------------------------------------------------------}
+
+writeEntry ::
+     HasFS IO h
+  -> HasBufFS IO h
+  -> MRun (FS.Handle h)
+  -> SerialisedKey
+  -> PrefixEntry (FS.Handle h)
+  -> IO ()
+writeEntry fs bfs writer key = \case
+    FullEntry entry -> do
+      -- simply add entry
+      MRun.addKeyOp fs writer key
+        =<< traverse resolveBlobRef entry
+    e@(OverflowEntry entryPrefix rawPage _ overflowPages) ->
+      assert (entryWouldFitInPage key entryPrefix) $
+      assert (prefixEntryInvariant e) $
+        if null entryPrefix  -- has blob?
+        then do
+          -- no blob, directly copy all pages as they are
+          MRun.addLargeSerialisedKeyOp fs writer key rawPage overflowPages
+        else do
+          -- has blob, we can't just copy the first page, fall back
+          -- we simply append the overflow pages to the value
+          MRun.addKeyOp fs writer key
+            =<< traverse resolveBlobRef (toFullEntry e)
+          -- TODO(optimise): This copies the overflow pages unnecessarily.
+          -- We could extend the MRun API to allow to either:
+          -- 1. write an Entry (containing the value prefix) + [RawOverflowPage]
+          -- 2. write a RawPage + SerialisedBlob + [RawOverflowPage], rewriting
+          --      the raw page's blob offset (slightly faster, but a bit hacky)
```

Comment:
I realised the MRun API is still not quite optimal, I feel tempted to just do option 2 now with something like `addLargeSerialisedKeyOp :: RawPage -> Maybe BlobOffset -> [RawOverflowPage] -> IO ()`.

Reply:
Leaving it for now, but might do it as part of a follow-up PR.